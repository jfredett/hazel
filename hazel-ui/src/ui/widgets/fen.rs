use ratatui::{buffer::Buffer, prelude::*, text::Text, widgets::Widget};

use crate::notation::ben::BEN;


#[derive(Debug, Default)]
#[allow(clippy::upper_case_acronyms)]
pub struct FEN {
    fen: BEN,
    style: Style,
    alignment: Alignment
}

impl FEN {
    pub fn new(fen: BEN) -> Self {
        Self {
            fen,
            style: Style::default(),
            alignment: Alignment::Left
        }
    }

    pub fn set_style(mut self, style: Style) -> Self {
        self.style = style;
        self
    }

    pub fn alignment(mut self, alignment: Alignment) -> Self {
        self.alignment = alignment;
        self
    }

    #[inline(always)]
    pub fn center(self) -> Self {
        self.alignment(Alignment::Center)
    }
}

impl Widget for &FEN {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let fenstring = format!("{}", self.fen);
        Text::styled(fenstring, self.style).alignment(self.alignment).render(area, buf);
    }
}

#[cfg(test)]
mod tests {
    use crate::constants::START_POSITION_FEN;
    use crate::board::simple::PieceBoard;

    use super::*;

    #[test]
    fn renders_empty_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let fen_widget = &FEN::new(BEN::empty());
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "8/8/8/8/8/8/8/8 w KQkq - 0 1                                    "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);
    }

    #[test]
    fn render_startpos_fen_correctly() {
        let rect = Rect::new(0, 0, 64, 1);
        let mut buffer = Buffer::empty(rect);
        buffer.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        let fen_widget = &FEN::new(BEN::new(START_POSITION_FEN));
        fen_widget.render(rect, &mut buffer);

        let mut expected = Buffer::with_lines(vec![
            "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1        "
        ]);
        expected.set_style(rect, Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(buffer, expected);

    }
}
