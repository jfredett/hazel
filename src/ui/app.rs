use std::collections::HashMap;
use std::fmt::Debug;

use ratatui::crossterm::event::{Event, KeyCode};
use ratatui::prelude::*;
use ratatui::widgets::{Block, Borders};

use crate::ui::widgets::game_section::board_section::fenwidget::FENWidget;
use crate::ui::widgets::game_section::board_section::boardwidget::BoardWidget;

use tracing::{debug, instrument};

use crate::uci::UCIMessage;
use crate::ui::model::entry::{Entry, stockfish};
use crate::engine::Engine;
use crate::ui::model::pieceboard::PieceBoard;

use super::widgets::tile::Tile;

enum Mode {
    Insert,
    Command
}

pub struct Hazel {
    flags: HashMap<String, bool>,
    entry: Entry,
    // UI
    mode: Mode,
    tile: Tile,
}

impl Debug for Hazel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Hazel")
            .field("flags", &self.flags)
            .finish()
    }
}

impl Hazel {
    #[instrument]
    pub fn new() -> Self {
        let mut s = Self {
            flags: HashMap::new(),
            mode: Mode::Command,
            entry: stockfish(),
            tile: Tile::new(),
        };

        s.entry.exec(UCIMessage::UCI);
        s.entry.exec(UCIMessage::IsReady);
        debug!("setting startpos");
        s.entry.exec(UCIMessage::Position("startpos moves d2d4".to_string(), vec![]));
        debug!("setting startpos done");

        // s.entry.boardstate.set_startpos();
        s.entry.boardstate = PieceBoard::from_fen("rnbqkbnr/pppppppp/8/8/3P4/8/PPP1PPPP/RNBQKBNR");

        return s;
    }

    #[instrument]
    pub fn handle_events(&mut self, event: Event) {
        if let Event::Key(key) = event {
            match self.mode {
                // Probably insert mode is just handled by the tile wholesale?
                Mode::Insert => {
                    match key.code {
                        KeyCode::Esc => {
                            self.mode = Mode::Command;
                        },
                        KeyCode::Char(c) => {
                            self.tile.handle_input(c);
                        },
                        KeyCode::Backspace => {
                            self.tile.handle_backspace();
                        },
                        KeyCode::Enter => {
                            self.tile.handle_enter();
                        },
                        _ => {
                        }
                    }
                },
                // Command mode will eventually select the tile you want/start new tiles, etc.
                Mode::Command => {
                    match key.code {
                        KeyCode::Char('i') => {
                            self.mode = Mode::Insert;
                        },
                        KeyCode::Char('q') => {
                            self.set_flag("exit", true);
                        },
                        _ => {
                        }
                    }
                }
            }
        }
    }

    #[instrument]
    pub fn set_flag(&mut self, flag: &str, value: bool) {
        self.flags.insert(flag.to_string(), value);
    }

    #[instrument]
    pub fn check_flag(&self, flag: &str) -> bool {
        match self.flags.get(flag) {
            Some(value) => *value,
            None => false
        }
    }

    #[instrument]
    pub fn fen_widget(&self) -> FENWidget {
        FENWidget::from(&self.entry)
    }

    #[instrument]
    pub fn board_widget(&self) -> BoardWidget {
        BoardWidget::from(&self.entry.boardstate)
    }

    #[instrument]
    pub fn input_widget(&self) -> Block {
        Block::default()
            .title("Input")
            .borders(Borders::ALL)
    }


    #[instrument]
    pub fn render(&mut self, frame: &mut Frame) {
        self.tile.set_state(self.entry.boardstate.clone());
        frame.render_widget(&self.tile, Rect::new(0,0,64,32));
    }
}

#[cfg(test)]
mod tests {
    use std::process::Termination;

    use backend::TestBackend;

    use super::*;

    #[test]
    fn placeholder() {
        let mut hazel = Hazel::new();

        let mut t = Terminal::new(TestBackend::new(64, 32)).unwrap();
        let _ = t.draw(|f| {
            hazel.render(f);
        });

        let mut expected = Buffer::with_lines(vec![
            "           Placeholder           ┌─────────────────────────────┐",
            "           Placeholder           │         Placeholder         │",
            "           Placeholder           │         Placeholder         │",
            "┌───────────────┐┌──────────────┐│         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder ││         Placeholder         │",
            "│  Placeholder  ││  Placeholder │└─────────────────────────────┘",
            "└───────────────┘└──────────────┘          Placeholder          ",
            "│                          Placeholder                         │",
            "┌──────────────────────────────────────────────────────────────┐",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "│                                                              │",
            "└──────────────────────────────────────────────────────────────┘",
            "$>                                                              ",
        ]);

        // Ignore style differences for now
        let mut actual = t.backend().buffer().clone();
        actual.set_style(Rect::new(0, 0, 64, 32), Style::default().fg(Color::White).bg(Color::Black));
        expected.set_style(Rect::new(0, 0, 64, 32), Style::default().fg(Color::White).bg(Color::Black));

        assert_eq!(actual, expected);
    }
}
