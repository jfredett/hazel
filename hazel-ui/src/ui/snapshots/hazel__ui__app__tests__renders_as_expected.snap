---
source: src/ui/app.rs
expression: t.backend().buffer().clone()
---
Buffer {
    area: Rect { x: 0, y: 0, width: 64, height: 32 },
    content: [
        "                   Tape: POS: 0x00000 (0/1), EOT: 0x0020        ",
        "                   ┌───────────────────────────────────────────┐",
        "                   │  Address  Instruction      Hash           │",
        "                   │>>0x0000                                   │",
        "a8 b8 c8 d8 e8 f8 g8 h8x0001                                   │",
        "                   │  0x0002                                   │",
        "a7 b7 c7 d7 e7 f7 g7 h7x0003                                   │",
        "                   │  0x0004                                   │",
        "a6 b6 c6 d6 e6 f6 g6 h6x0005                                   │",
        "                   │  0x0006                                   │",
        "a5 b5 c5 d5 e5 f5 g5 h5x0007                                   │",
        "                   │  0x0008                                   │",
        "a4 b4 c4 d4 e4 f4 g4 h4x0009                                   │",
        "8/8/8/8/8/8/8/8 w K│  0x000A                                   │",
        "a3 b3 c3 d3 e3 f3 g3─h3────────────────────────────────────────┘",
        "                   Footer here                                  ",
        "┌Tui Ta┐┌Tui Log [log=0.0/s]───────────────────────────────────┐",
        "│      ││                                                      │",
        "│1 b1 c││d1 e1 f1 g1 h1                                        │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "│      ││                                                      │",
        "└──────┘└──────────────────────────────────────────────────────┘",
        "┌──────────────────────────────────────────────────────────────┐",
        "│                                                              │",
        "│                                                              │",
        "│                                                              │",
        "└──────────────────────────────────────────────────────────────┘",
        "$>                                                              ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 19, y: 1, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 19, y: 2, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 2, fg: White, bg: Reset, underline: Reset, modifier: BOLD,
        x: 63, y: 2, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 3, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 2, y: 3, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 5, y: 3, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 7, y: 3, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 10, y: 3, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 12, y: 3, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 14, y: 3, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 17, y: 3, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 19, y: 3, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 3, fg: White, bg: Reset, underline: Reset, modifier: REVERSED,
        x: 63, y: 3, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 2, y: 4, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 3, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 5, y: 4, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 6, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 7, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 8, y: 4, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 9, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 10, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 11, y: 4, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 12, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 14, y: 4, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 15, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 17, y: 4, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 18, y: 4, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 19, y: 4, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 4, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 4, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 4, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 5, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 2, y: 5, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 5, y: 5, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 7, y: 5, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 10, y: 5, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 12, y: 5, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 14, y: 5, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 17, y: 5, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 19, y: 5, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 2, y: 6, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 3, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 5, y: 6, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 6, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 7, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 8, y: 6, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 9, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 10, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 11, y: 6, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 12, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 14, y: 6, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 15, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 17, y: 6, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 18, y: 6, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 19, y: 6, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 6, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 6, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 6, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 7, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 2, y: 7, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 5, y: 7, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 7, y: 7, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 10, y: 7, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 12, y: 7, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 14, y: 7, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 17, y: 7, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 19, y: 7, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 2, y: 8, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 3, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 5, y: 8, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 6, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 7, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 8, y: 8, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 9, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 10, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 11, y: 8, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 12, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 14, y: 8, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 15, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 17, y: 8, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 18, y: 8, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 19, y: 8, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 8, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 8, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 8, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 9, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 2, y: 9, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 5, y: 9, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 7, y: 9, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 10, y: 9, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 12, y: 9, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 14, y: 9, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 17, y: 9, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 19, y: 9, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 2, y: 10, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 3, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 5, y: 10, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 6, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 7, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 8, y: 10, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 9, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 10, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 11, y: 10, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 12, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 14, y: 10, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 15, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 17, y: 10, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 18, y: 10, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 19, y: 10, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 10, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 10, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 10, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 11, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 2, y: 11, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 5, y: 11, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 7, y: 11, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 10, y: 11, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 12, y: 11, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 14, y: 11, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 17, y: 11, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 19, y: 11, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 2, y: 12, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 3, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 5, y: 12, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 6, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 7, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 8, y: 12, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 9, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 10, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 11, y: 12, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 12, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 14, y: 12, fg: Black, bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 15, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(160, 160, 160), underline: Reset, modifier: NONE,
        x: 17, y: 12, fg: White, bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 18, y: 12, fg: Rgb(101, 101, 101), bg: Rgb(96, 96, 96), underline: Reset, modifier: NONE,
        x: 19, y: 12, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 12, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 12, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 12, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 13, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 19, y: 13, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 2, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 3, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 5, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 6, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 11, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 14, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 15, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 14, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 14, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 14, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 14, fg: White, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 15, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 2, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 3, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 5, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 6, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 11, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 14, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 15, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 16, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 16, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 2, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 3, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 5, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 6, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 11, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 12, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 14, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 15, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 17, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 18, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 20, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 21, y: 18, fg: Rgb(101, 101, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 23, y: 18, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 0, y: 26, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 27, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 63, y: 27, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 28, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 63, y: 28, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 1, y: 29, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 63, y: 29, fg: White, bg: Black, underline: Reset, modifier: NONE,
        x: 0, y: 31, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
